dkg/pedersen/handler.go

```diff
@@ -0,0 +1,386 @@
+var suite = suites.MustFind("Ed25519")
+
+// Handler represents the RPC executed on each node
+//
+// - implements mino.Handler
+type Handler struct {
+	mino.UnsupportedHandler
+	sync.RWMutex
+	af        mino.AddressFactory
+	dkg       *pedersen.DistKeyGenerator
+	pubKeys   []kyber.Point
+	privKey   kyber.Scalar
+	me        mino.Address
+	suite     suites.Suite
+	privShare *share.PriShare
+}
+
+// NewHandler creates a new handler
+func NewHandler(pubKeys []kyber.Point, privKey kyber.Scalar,
+	af mino.AddressFactory, me mino.Address, suite suites.Suite) *Handler {
+
+	return &Handler{
+		af:      af,
+		pubKeys: pubKeys,
+		privKey: privKey,
+		me:      me,
+		suite:   suite,
+	}
+}
+
+// Stream implements mino.Handler. It allows one to stream messages to the
+// players.
+func (h *Handler) Stream(out mino.Sender, in mino.Receiver) error {
+	// Note: one should never assume any synchronous properties on the messages.
+	// For example we can not expect to receive the start message from the
+	// initiator of the DKG protocol first because some node could have received
+	// this start message earlier than us, start their DKG work by sending
+	// messages to the other nodes, and then we might get their messages before
+	// the start message.
+
+	from, msg, err := in.Recv(context.Background())
+	if err != nil {
+		return xerrors.Errorf("failed to receive: %v", err)
+	}
+
+	// We expect a Start message or a decrypt request at first, but we might
+	// receive other messages in the meantime, like a Deal.
+	switch msg := msg.(type) {
+
+	case *Start:
+		err := h.start(msg, []*Deal{}, from, out, in)
+		if err != nil {
+			return xerrors.Errorf("failed to start: %v", err)
+		}
+
+	case *DecryptRequest:
+		// TODO: check if started before
+		K := h.suite.Point()
+		err := K.UnmarshalBinary(msg.K)
+		if err != nil {
+			return xerrors.Errorf("failed to unmarshal K: %v", err)
+		}
+
+		C := h.suite.Point()
+		err = C.UnmarshalBinary(msg.C)
+		if err != nil {
+			return xerrors.Errorf("failed to unmarshal C: %v", err)
+		}
+
+		h.RLock()
+		S := suite.Point().Mul(h.privShare.V, K)
+		h.RUnlock()
+
+		partial := suite.Point().Sub(C, S)
+
+		VBuf, err := partial.MarshalBinary()
+		if err != nil {
+			return xerrors.Errorf("failed to marshal the partial: %v", err)
+		}
+
+		h.RLock()
+		decryptReply := &DecryptReply{
+			V: VBuf,
+			// TODO: check if using the private index is the same as the public
+			// index.
+			I: int64(h.privShare.I),
+		}
+		h.RUnlock()
+
+		errs := out.Send(decryptReply, from)
+		err, more := <-errs
+		if more {
+			return xerrors.Errorf("got an error while sending the decrypt "+
+				"reply: %v", err)
+		}
+
+	case *Deal:
+		// This is a special case where a DKG started, some nodes received the
+		// start signal and started sending their deals but we have not yet
+		// received our start signal. In this case we will collect the Deals
+		// while waiting for the start signal.
+		deals := []*Deal{msg}
+		for {
```

Comment:
You're essentially doing the same thing so you can probably for loop the whole select and remove this one.

Reply:
You just need to handle the _Decrypt_ if it hasn't started but you have to do it anyway.